tests/test-output-flex-way-change.cpp

```diff
@@ -0,0 +1,274 @@
+                              "way_id = 10 AND rel_ids = '{30}'"));
+
+    options.append = true;
+
+    REQUIRE_NOTHROW(db.run_import(
+        options, "r30 v2 dV Tt=ag Mw10@,w11@,w12@mark,w13@,w14@mark"));
+
+    CHECK(1 == conn.get_count("osm2pgsql_test_t1"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_t2"));
+    CHECK(3 == conn.get_count("osm2pgsql_test_tboth"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_tboth",
+                              "way_id = 10 AND rel_ids IS NULL"));
+}
+
+TEST_CASE("change tboth without stage2 to tboth with stage2")
+{
+    options_t options = testing::opt_t().slim().flex(conf_file);
+
+    testing::data_t data{tdata};
+    data.add({"w10 v1 dV Ttboth=yes Nn10,n11",
+              "r30 v1 dV Tt=ag Mw10@,w11@,w12@mark,w13@,w14@mark"});
+
+    REQUIRE_NOTHROW(db.run_import(options, data()));
+
+    auto conn = db.db().connect();
+
+    CHECK(1 == conn.get_count("osm2pgsql_test_t1"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_t2"));
+    CHECK(3 == conn.get_count("osm2pgsql_test_tboth"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_tboth",
+                              "way_id = 10 AND rel_ids IS NULL"));
+
+    options.append = true;
+
+    REQUIRE_NOTHROW(db.run_import(
+        options, "r30 v2 dV Tt=ag Mw10@mark,w11@,w12@mark,w13@,w14@mark"));
+
+    CHECK(1 == conn.get_count("osm2pgsql_test_t1"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_t2"));
+    CHECK(3 == conn.get_count("osm2pgsql_test_tboth"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_tboth",
+                              "way_id = 10 AND rel_ids = '{30}'"));
+}
+
+TEST_CASE("change valid geom to invalid geom")
+{
+    options_t options = testing::opt_t().slim().flex(conf_file);
+
+    testing::data_t data{tdata};
+    data.add({"w10 v1 dV Tt1=yes,t2=yes,tboth=yes Nn10,n11",
+              "r30 v1 dV Tt=ag Mw10@mark,w11@,w12@mark,w13@,w14@mark"});
+
+    REQUIRE_NOTHROW(db.run_import(options, data()));
+
+    auto conn = db.db().connect();
+
+    CHECK(2 == conn.get_count("osm2pgsql_test_t1"));
+    CHECK(2 == conn.get_count("osm2pgsql_test_t2"));
+    CHECK(3 == conn.get_count("osm2pgsql_test_tboth"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_tboth",
+                              "way_id = 10 AND rel_ids = '{30}'"));
+
+    options.append = true;
+
+    SECTION("change node list to make way invalid")
+    {
+        REQUIRE_NOTHROW(
+            db.run_import(options, "w10 v2 dV Tt1=yes,t2=yes,tboth=yes Nn10"));
+    }
+
+    SECTION("change node to make way invalid (n11 same location as n10)")
+    {
+        REQUIRE_NOTHROW(db.run_import(options, "n11 v2 dV x10.0 y10.0"));
+    }
+
+    CHECK(1 == conn.get_count("osm2pgsql_test_t1"));
+    CHECK(1 == conn.get_count("osm2pgsql_test_t2"));
+    CHECK(2 == conn.get_count("osm2pgsql_test_tboth"));
+    CHECK(0 == conn.get_count("osm2pgsql_test_tboth",
+                              "way_id = 10 AND rel_ids = '{30}'"));
```

Comment:
Test for "way_id = 10" only, it's broader.

Reply:
fixed